sdk/src/wallet/operations/transaction/high_level/staking/begin.rs

```diff
@@ -0,0 +1,99 @@
+pub struct BeginStakingParams {
+    /// The account id which will become a validator.
+    pub account_id: AccountId,
+    /// The amount of tokens to stake.
+    pub staked_amount: u64,
+    /// The fixed cost of the validator, which it receives as part of its Mana rewards.
+    pub fixed_cost: u64,
+    /// The staking period (in epochs). Will default to the staking unbonding period.
+    pub staking_period: Option<u32>,
+}
+
+impl<S: 'static + SecretManage> Wallet<S>
+where
+    crate::wallet::Error: From<S::Error>,
+    crate::client::Error: From<S::Error>,
+{
+    pub async fn begin_staking(
+        &self,
+        params: BeginStakingParams,
+        options: impl Into<Option<TransactionOptions>> + Send,
+    ) -> crate::wallet::Result<TransactionWithMetadata> {
+        let options = options.into();
+        let prepared = self.prepare_begin_staking(params, options.clone()).await?;
+
+        self.sign_and_submit_transaction(prepared, None, options).await
+    }
+
+    /// Prepares the transaction for [Wallet::begin_staking()].
+    pub async fn prepare_begin_staking(
+        &self,
+        params: BeginStakingParams,
+        options: impl Into<Option<TransactionOptions>> + Send,
+    ) -> crate::wallet::Result<PreparedTransactionData> {
+        log::debug!("[TRANSACTION] prepare_begin_staking");
+
+        let account_id = params.account_id;
+        let account_output_data = self
+            .data()
+            .await
+            .unspent_account_output(&account_id)
+            .cloned()
+            .ok_or_else(|| crate::wallet::Error::AccountNotFound)?;
+
+        if account_output_data
+            .output
+            .features()
+            .map_or(false, |f| f.staking().is_some())
+        {
+            return Err(crate::wallet::Error::StakingFailed(format!(
+                "account id {account_id} already has a staking feature"
+            )));
+        }
+
+        let protocol_parameters = self.client().get_protocol_parameters().await?;
+
+        if let Some(staking_period) = params.staking_period {
+            if staking_period < protocol_parameters.staking_unbonding_period() {
+                return Err(crate::wallet::Error::StakingFailed(format!(
+                    "staking period {staking_period} is less than the minimum {}",
+                    protocol_parameters.staking_unbonding_period()
+                )));
+            }
+        }
+
+        let slot_commitment_id = self.client().get_issuance().await?.latest_commitment.id();
+        let start_epoch = protocol_parameters.epoch_index_of(protocol_parameters.past_bounded_slot(slot_commitment_id));
+
+        let output = AccountOutputBuilder::from(account_output_data.output.as_account())
+            .with_account_id(account_id)
+            .add_feature(StakingFeature::new(
+                params.staked_amount,
+                params.fixed_cost,
+                start_epoch,
+                EpochIndex(u32::MAX),
```

Comment:
Should this only be the default if no `staking_period` has been provided?

Reply:
ahhhh dang it I'm all over the place today